### PR TITLE
ticket-176

### PR DESCRIPTION
**Pull request by mkante**
Fix traceID restore problem caused by CLS

### CODE REVIEW

**Comment by eirslett** [comment]
@adriancole do you have time to look at this issue/PR? I'm not 100 % sure it's the correct solution to the problem, but at least it looks like an improvement! I'm a bit concerned it could cause regressions or changed behavior in other cases. But unfortunately, I don't have time to look closely at it myself.

**Comment by mkante** [comment]
I see `continuous-integration/appveyor/pr` failure from Kafka,  I don't think it's related to my PR. Once  have an environment setup with kafka, I will look into it.

**Comment by adriancole** [comment]
thank you. I will make time to look at this, as it is important. Thanks for the work here @mkante!

File: packages/zipkin-instrumentation-memcached/src/zipkinClient.js [hunk 1]
```diff
@@ -17,6 +17,7 @@ module.exports = function zipkinClient(
         }));
         tracer.recordAnnotation(new Annotation.ClientRecv());
       });
+      tracer.setId(originalID);
```
Comment (adriancole) [inline packages/zipkin-instrumentation-memcached/src/zipkinClient.js]: so this line implies a leak I guess

Comment (mkante) [reply packages/zipkin-instrumentation-memcached/src/zipkinClient.js]: Correct

Comment (mkante) [reply packages/zipkin-instrumentation-memcached/src/zipkinClient.js]: I shared small service on [issue 172](https://github.com/openzipkin/zipkin-js/issues/176) to prove my point

Comment (adriancole) [reply packages/zipkin-instrumentation-memcached/src/zipkinClient.js]: visual check, all calls to setId happen in scoped blocks, so this guard is indeed interesting

Comment (adriancole) [reply packages/zipkin-instrumentation-memcached/src/zipkinClient.js]: the test passes with this commented out.. however the test uses explicit context, not CLS. CLS is the scenario of your example I think...

Comment (adriancole) [reply packages/zipkin-instrumentation-memcached/src/zipkinClient.js]: this also passes with CLS..

**Comment by adriancole** [comment]
after #178 I think the below patch might make things slightly more explicit. Only thing I have concern with is that since the test breaks before and after.. we should write a test that breaks before :P are you able to help with that?

```diff
-  function mkZipkinCallback(callback, id, originalID) {
+  function mkZipkinCallback(callback, id) {
+    const originalId = tracer.id;
     return function zipkinCallback(...args) {
       tracer.letId(id, () => {
         // TODO: parse host and port from details in callback
@@ -16,8 +17,10 @@ module.exports = function zipkinClient(
         }));
         tracer.recordAnnotation(new Annotation.ClientRecv());
       });
-      tracer.setId(originalID);
-      callback.apply(this, args);
+      // callback runs after the client request, so let's restore the former ID
+      tracer.letId(originalId, () => {
+        callback.apply(this, args);
+      });
     };
   }
```

**Comment by adriancole** [comment]
meant to say the test passes before and after the change to src/

**Comment by adriancole** [comment]
probably to get the test to fail, we need to manually manipulate the trace ID in one of the callbacks or somehow get the test to run twice

**Comment by mkante** [comment]
If I understand you want to merge #178 first, then I can apply the code above? That wont be problem. Do you know if #178 is going in any time soon?

**Comment by adriancole** [comment]
178 just needs a review.. care to?

**Comment by mkante** [comment]
No problem

**Pull request merged**